fsm/builder.go

```diff
@@ -43,7 +48,11 @@ func (b *Builder) SetTransitions(transitions []Transition) {
 }
 
 // Build constructs a new FSM
-func (b *Builder) Build() (*FSM, error) {
+func (b *Builder) Build(stater Stater) (*FSM, error) {
```

Comment:
Now this is probably more like an `Apply` instead of a `Build` but I just kept with the existing taxonomy.

Reply:
Or `Create()`.  I haven't verified this (there's low reuse across builder instances), but ideally `Build()` copies all of its members so the Builder can be reused.  At the time my thinking was, "builders build or create things."

Reply:
> ideally `Build()` copies all of its members so the Builder can be reused

I definitely like the Builder pattern here but reuse will be tough when each callback is only defined once for a single guard/callback. You need some object at runtime to run conditions against. That was sort of the point of this first PR, not so much serialization yet.

I'm not a hard advocate for what I introduced but I needed this sort of `Stater` object abstraction at each callback. It felt like the utility of this was defining a bunch of guards/callbacks (a "rule set") up front then, at runtime, building an FSM for each object that you need to derive state of. This FSM doesn't have a `Run()` func so it isn't going to be used like a lexical parser would, where you step through each state with a main `switch` statement. This will still need something external to cycle through each `Transition()` call, why not pass that external object into the FSM so it is available for each callback.

Also, we might need something that allows us to set the state from an external source than run the `Enter` callback to validate that state prior to moving on to the next `Transition`. We don't have this yet but passing an object into the FSM would get us closer.

Reply:
Offline discussion was held to handle state transitions in gRPC handlers.